util/extsort/disk_sorter.go

```diff
@@ -15,37 +15,654 @@
+	c.lastKey = append(c.lastKey[:0], key.UserKey...)
+	if c.curSize >= c.bucketSize {
+		c.addBucket()
+	}
+	return nil
+}
+
+func (c *kvStatsCollector) addBucket() {
+	c.buckets = append(c.buckets, kvStatsBucket{
+		Size:       c.curSize,
+		UpperBound: slices.Clone(c.lastKey),
+	})
+	c.curSize = 0
+}
+
+func (c *kvStatsCollector) Finish(userProps map[string]string) error {
+	if c.curSize > 0 {
+		c.addBucket()
+	}
+	stats := kvStats{Histogram: c.buckets}
+	data, err := json.Marshal(&stats)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	userProps[kvStatsPropKey] = string(data)
+	return nil
+}
+
+func (*kvStatsCollector) Name() string {
+	return kvStatsPropKey
+}
+
+func makeFilename(fs vfs.FS, dirname string, fileNum int) string {
+	return fs.PathJoin(dirname, fmt.Sprintf("%06d%s", fileNum, sstFileSuffix))
+}
+
+func parseFilename(fs vfs.FS, filename string) (fileNum int, ok bool) {
+	filename = fs.PathBase(filename)
+	if !strings.HasSuffix(filename, sstFileSuffix) {
+		return 0, false
+	}
+	v, err := strconv.ParseInt(filename[:len(filename)-len(sstFileSuffix)], 10, 64)
+	if err != nil {
+		return 0, false
+	}
+	return int(v), true
+}
+
+type sstWriter struct {
+	w         *sstable.Writer
+	fs        vfs.FS
+	fileNum   int
+	tmpPath   string
+	destPath  string
+	err       error
+	onSuccess func(meta *fileMetadata)
+}
+
+func newSSTWriter(
+	fs vfs.FS,
+	dirname string,
+	fileNum int,
+	kvStatsBucketSize int,
+	onSuccess func(meta *fileMetadata),
+) (*sstWriter, error) {
+	destPath := makeFilename(fs, dirname, fileNum)
+	tmpPath := destPath + tmpFileSuffix
+	f, err := fs.Create(tmpPath)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	w := sstable.NewWriter(f, sstable.WriterOptions{
+		TablePropertyCollectors: []func() sstable.TablePropertyCollector{
+			func() sstable.TablePropertyCollector {
+				return newKVStatsCollector(kvStatsBucketSize)
+			},
+		},
+	})
+
+	sw := &sstWriter{
+		w:         w,
+		fs:        fs,
+		fileNum:   fileNum,
+		tmpPath:   tmpPath,
+		destPath:  destPath,
+		onSuccess: onSuccess,
+	}
+	return sw, nil
+}
+
+func (sw *sstWriter) Set(key, value []byte) error {
+	sw.err = sw.w.Set(key, value)
+	return errors.Trace(sw.err)
+}
+
+func (sw *sstWriter) Close() (retErr error) {
+	defer func() {
+		if retErr == nil {
+			retErr = sw.fs.Rename(sw.tmpPath, sw.destPath)
+		} else {
+			_ = sw.fs.Remove(sw.tmpPath)
+		}
+	}()
+	closeErr := sw.w.Close()
+	if err := goerrors.Join(sw.err, closeErr); err != nil {
+		return errors.Trace(err)
+	}
+
+	if sw.onSuccess != nil {
+		writerMeta, err := sw.w.Metadata()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		meta := &fileMetadata{
+			fileNum: sw.fileNum,
+		}
+		smallest := writerMeta.Smallest(pebble.DefaultComparer.Compare)
+		largest := writerMeta.Largest(pebble.DefaultComparer.Compare)
+		meta.startKey = slices.Clone(smallest.UserKey)
+		meta.lastKey = slices.Clone(largest.UserKey)
+		meta.endKey = append(meta.lastKey, 0)
```

Comment:
why add 0 to it, can we comment before it?

Reply:
`endKey` is exclusive. See the comment in `fileMetadata`.

Reply:
ok, but we still need a comment here to explain why add 0

Reply:
https://github.com/pingcap/tidb/pull/44185/commits/e843887215eb44107a5d05d30d10fc1f57acb144